core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java

```diff
@@ -0,0 +1,191 @@
+    for (int i = 0; i < components.length; i++) {
+      if (i > 0 && i <= mountPointBase) {
+        // Do not process components before the base of the mount point.
+        // However, process the first component, since that will be the actual mount point base.
+        continue;
+      }
+      String component = components[i];
+      uri = uri.join(component);
+      String uriPath = uri.getPath();
+      try {
+        if (ufs.exists(uri.toString())) {
+          // This ufs path exists. Remove the cache entry.
+          cache.remove(uriPath);
+        } else {
+          // This is the first ufs path which does not exist. Add it to the cache.
+          cache.add(uriPath);
+
+          // Remove cache entries which has this non-existing directory as a prefix. This is not for
+          // correctness, but to "compress" information.
+          String dirPath = uriPath + "/";
+          Iterator<String> it = cache.tailSet(dirPath).iterator();
+          while (it.hasNext()) {
+            String existingPath = it.next();
+            if (existingPath.startsWith(dirPath)) {
+              // An existing cache entry has the non-existing path as a prefix. Remove the entry,
+              // since the non-existing path ancestor implies the descendant does not exist.
+              it.remove();
+            }
+          }
+          // The first non-existing path was found, so further traversal is unnecessary.
+          break;
+        }
+      } catch (IOException e) {
+        throw new InvalidPathException(ExceptionMessage.PATH_DOES_NOT_EXIST.getMessage(path), e);
+      }
+    }
+  }
+
+  /**
+   * Removes an absent path from the cache.
+   *
+   * @param path the path to remove from the cache
+   * @throws InvalidPathException if the path is invalid
+   */
+  public void removeAbsentPath(AlluxioURI path) throws InvalidPathException {
+    MountTable.Resolution resolution = mMountTable.resolve(path);
+    AlluxioURI ufsUri = resolution.getUri();
+    ConcurrentSkipListSet<String> cache = getCache(resolution.getMountId());
+
+    String[] components = PathUtils.getPathComponents(ufsUri.getPath());
+
+    // create a ufs uri of the root.
+    AlluxioURI uri =
+        new AlluxioURI(ufsUri.getScheme(), ufsUri.getAuthority(), "/", ufsUri.getQueryMap());
+
+    // Traverse through the ufs path components, staring from the root.
+    for (String component : components) {
+      uri = uri.join(component);
+      String uriPath = uri.getPath();
+      // This ufs path exists. Remove the cache entry.
+      cache.remove(uriPath);
+    }
+  }
+
+  /**
+   * Returns true if the given path is absent, according to this cache. A path is absent if one of
+   * its ancestors is absent.
+   *
+   * @param path the path to check
+   * @return true if the path is absent according to the cache
+   * @throws InvalidPathException if the path is invalid
+   */
+  public boolean isAbsent(AlluxioURI path) throws InvalidPathException {
```

Comment:
I originally thought `contains` would have been confusing, since if this contains a path, it means it doesn't exist in ufs. Do you think it won't be confusing?